src/app/tabs/camera.cpp

```diff
@@ -1,135 +1,361 @@
+    input->setFont(Theme::font_18);
+    input->setAlignment(Qt::AlignCenter);
+    connect(input, &QLineEdit::textEdited, [this](QString text) { this->config->set_cam_network_url(text); });
+    connect(input, &QLineEdit::returnPressed, this, &CameraTab::connect_network_stream);
+
+    layout->addStretch(1);
+    layout->addWidget(input, 4);
+    layout->addStretch(1);
+
+    return widget;
+}
+
+void CameraTab::update_network_status(QMediaPlayer::MediaStatus media_status)
+{
+    qInfo() << "camera status changed to: " << media_status;
+
+    switch (media_status) {
+        case QMediaPlayer::LoadingMedia:
+        case QMediaPlayer::LoadedMedia:
+        case QMediaPlayer::BufferedMedia:
+            this->status->setText("connecting...");
+            break;
+        default:
+            this->status->setText("connection failed");
+            emit disconnected();
+            break;
+    }
+}
+
+void CameraTab::connect_network_stream()
+{
+    connect(this->player, &QMediaPlayer::mediaStatusChanged,
+            [this](QMediaPlayer::MediaStatus media_status) { this->update_network_status(media_status); });
+    connect(this->player, QOverload<>::of(&QMediaPlayer::metaDataChanged), [this]() { emit connected_network(); });
+    qInfo() << "playing stream: " << this->config->get_cam_network_url();
+    this->player->setMedia(QUrl(this->config->get_cam_network_url()));
+    this->player->play();
+}
+
+void CameraTab::connect_local_stream()
+{
+    if (this->local_cam != nullptr) {
+        delete this->local_cam;
+        this->local_cam = nullptr;
+    }
+
+    const QString& local = this->config->get_cam_local_device();
+    if (!this->local_cam_available(local)) {
+        this->status->setText("Camera unavailable");
+        return;
+    }
+
+    qDebug() << "Connecting to local cam " << local;
+    this->local_cam = new QCamera(local.toUtf8(), this);
+    this->local_cam->setViewfinder(this->local_video_widget);
+    connect(this->local_cam, &QCamera::statusChanged, this, &CameraTab::update_local_status);
+    this->local_cam->start();
+}
+
+void CameraTab::choose_video_resolution()
+{
+    QSize window_size = this->size();
+    QCameraImageCapture imageCapture(this->local_cam);
+    int min_gap = 10000, xgap, ygap;
+    QSize max_fit;
+    for (auto const& resolution : imageCapture.supportedResolutions()) {
+        xgap = window_size.width() - resolution.width();
+        ygap = window_size.height() - resolution.height();
+        if (xgap >= 0 && ygap >= 0 && xgap+ygap < min_gap) {
+            min_gap = xgap + ygap;
+            max_fit = resolution;
+        }
+    }
+    if (max_fit.isValid()) {
+        qDebug() << "Local cam auto resolution" << max_fit << "to fit in" << window_size;
+        this->local_cam_settings.setResolution(max_fit);
+    }
+    else
+        qDebug() << "No suitable resolutions found to fit in" << window_size;
+    //    this->local_cam_settings.setPixelFormat(QVideoFrame::Format_YV12);
```

Comment:
This is for @robert5974 to uncomment on his machine until we understand why his machine is mismatching formats.